source/structuring.rst

```diff
@@ -14,402 +14,545 @@ functions is better than copying-and-pasting duplicate bits of code
+    {
+      "name": "Elizabeth",
+      "children": [
+        {
+          "name": "Charles",
+          "children": [
+            {
+              "name": "William",
+              "children": [
+                { "name": "George" },
+                { "name": "Charlotte" }
+              ]
+            },
+            {
+              "name": "Harry"
+            }
+          ]
+        }
+      ]
+    }
 
-.. note::
+Above, we created a schema that refers to itself, effectively creating
+a "loop" in the validator, which is both allowed and useful. Note,
+however, that a ``$ref`` referring to another ``$ref`` could cause
+an infinite loop in the resolver, and is explicitly disallowed.
 
-    This functionality isn't currently supported by the Python ``jsonschema``
-    library.
+.. schema_example::
 
-Extending
----------
+    {
+      "definitions": {
+        "alice": { "$ref": "#/definitions/bob" },
+        "bob": { "$ref": "#/definitions/alice" }
+      }
+    }
 
-The power of ``$ref`` really shines when it is used with the
-combining keywords ``allOf``, ``anyOf`` and ``oneOf`` (see
-:ref:`combining`).
+.. index::
+   single: bundling
+   single: $id
+   single: structuring; bundling; $id
 
-Let's say that for a shipping address, we want to know whether the
-address is a residential or business address, because the shipping
-method used may depend on that.  For a billing address, we don't
-want to store that information, because it's not applicable.
+.. _bundling:
 
-To handle this, we'll update our definition of shipping address::
+Bundling
+--------
 
-    "shipping_address": { "$ref": "#/definitions/address" }
+Working with multiple schema documents is convenient for development,
+but it is often more convenient for distribution to bundle all of your
+schemas into a single schema document. This can be done using the
+``$id`` keyword.
 
-to instead use an ``allOf`` keyword entry combining both the core
-address schema definition and an extra schema snippet for the address
-type::
+.. draft_specific::
 
-    "shipping_address": {
-      "allOf": [
-        // Here, we include our "core" address schema...
-        { "$ref": "#/definitions/address" },
+    --Draft 4
+    In Draft 4, ``$id`` is just ``id`` (without the dollar sign).
 
-        // ...and then extend it with stuff specific to a shipping
-        // address
-        { "properties": {
-            "type": { "enum": [ "residential", "business" ] }
-          },
-          "required": ["type"]
-        }
-      ]
-    }
+When ``$id`` is used in a subschema, it creates a new identifier that
+any references in that subschema and any below it will resolve against
+instead of the schema document's identifier. The new identifier is the
+value of ``$id`` resolved against the identifier of the schema it
+appears in.
 
-Tying this all together,
+This example shows the customer schema example and the address schema
+example bundled into a single schema document.
```

Comment:
```suggestion
example bundled into a Compound Schema Document.
```

Reply:
You can't technically make a Compound Schema in draft-07.